components/authz-service/server/v2/policy_refresher.go

```diff
@@ -0,0 +1,276 @@
+package v2
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/components/authz-service/engine"
+	storage "github.com/chef/automate/components/authz-service/storage/v2"
+	"github.com/chef/automate/lib/logger"
+	"google.golang.org/grpc/metadata"
+)
+
+var ErrMessageBoxFull = errors.New("Message box full")
+
+type PolicyRefresher interface {
+	Refresh(ctx context.Context) error
+	RefreshAsync() error
+}
+
+type policyRefersher struct {
+	log                      logger.Logger
+	store                    storage.Storage
+	engine                   engine.V2Writer
+	refreshRequests          chan policyRefresherMessageRefresh
+	antiEntropyTimerDuration time.Duration
+}
+
+type policyRefresherMessageRefresh struct {
+	ctx    context.Context
+	status chan error
+}
+
+func (m *policyRefresherMessageRefresh) Respond(err error) {
+	if m.status != nil {
+		select {
+		case m.status <- err:
+		default:
+		}
+		close(m.status)
+	}
+}
+
+func (m *policyRefresherMessageRefresh) Err() error {
+	return <-m.status
+}
+
+func NewPolicyRefresher(ctx context.Context, log logger.Logger, store storage.Storage,
+	engine engine.V2Writer) PolicyRefresher {
+	refresher := &policyRefersher{
+		log:                      log,
+		store:                    store,
+		engine:                   engine,
+		refreshRequests:          make(chan policyRefresherMessageRefresh, 1),
+		antiEntropyTimerDuration: 10 * time.Second,
+	}
+	go refresher.run(ctx)
+	return refresher
+}
+
+func (refresher *policyRefersher) run(ctx context.Context) {
+	lastPolicyID := ""
+	antiEntropyTimer := time.NewTimer(refresher.antiEntropyTimerDuration)
+RUNLOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			refresher.log.WithError(ctx.Err()).Info("Policy refresher exiting")
+			break RUNLOOP
+		case m := <-refresher.refreshRequests:
+			var err error
+			lastPolicyID, err = refresher.refresh(m.ctx, lastPolicyID)
+			m.Respond(err)
+			if !antiEntropyTimer.Stop() {
+				<-antiEntropyTimer.C
+			}
+		case <-antiEntropyTimer.C:
+			var err error
+			lastPolicyID, err = refresher.refresh(ctx, lastPolicyID)
+			if err != nil {
+				refresher.log.WithError(err).Warn("Anti-entropy refresh failed")
+			}
+		}
+
+		antiEntropyTimer.Reset(refresher.antiEntropyTimerDuration)
+	}
+	close(refresher.refreshRequests)
+}
+
+func (refresher *policyRefersher) refresh(ctx context.Context, lastPolicyID string) (string, error) {
+	curPolicyID, err := refresher.store.GetPolicyChangeID(ctx)
+	if err != nil {
+		refresher.log.WithError(err).Warn("Failed to get current policy change ID")
+		return lastPolicyID, err
```

Comment:
So the first returned value is _OK to use_ even if there's a non-nil error? That's a bit against conventions, I believe. 🤔

Reply:
not sure i understand. The intent is that if there is an error, the change id doesn't change